src/libstd/io/stdio.rs

```diff
@@ -175,13 +175,17 @@ impl Read for Stdin {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<'a> Read for StdinLock<'a> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        self.inner.read(buf)
+        self.inner.borrow_mut().read(buf)
     }
 }
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<'a> BufRead for StdinLock<'a> {
-    fn fill_buf(&mut self) -> io::Result<&[u8]> { self.inner.fill_buf() }
-    fn consume(&mut self, n: usize) { self.inner.consume(n) }
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        self.inner.borrow_mut().fill_buf()
```

Comment:
Hm this is unfortunately a little tricky here, the nested `fill_buf()` return value is only valid for the length of time as the `borrow_mut()`, which isn't the length of time of the outer caller unfortunately. I think that the `StdinLock` structure may want to be redefined as:

```rust
pub struct StdinLock<'a> {
    lock: ReentrantMutexGuard<'a, RefCell<T>>,
    borrow: RefMut<'a, T>,
}
```

I think then you can call `self.borrow.fill_buf()` and it should work out?